src/palace/manager/api/odl/auth.py

```diff
@@ -0,0 +1,147 @@
+import json
+from abc import ABC, abstractmethod
+from collections.abc import Mapping
+from typing import Any
+
+from requests import Response
+
+from palace.manager.api.odl.settings import OPDS2AuthType
+from palace.manager.core.exceptions import IntegrationException
+from palace.manager.util.http import HTTP, BearerAuth
+
+
+class ODLAuthenticatedGet(ABC):
+    AUTH_DOC_CONTENT_TYPES = [
+        "application/opds-authentication+json",
+        "application/vnd.opds.authentication.v1.0+json",
+    ]
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._session_token: str | None = None
+
+    @property
+    @abstractmethod
+    def _username(self) -> str:
+        ...
+
+    @property
+    @abstractmethod
+    def _password(self) -> str:
+        ...
+
+    @property
+    @abstractmethod
+    def _auth_type(self) -> OPDS2AuthType:
+        ...
+
+    def _basic_auth_get(
+        self, url: str, headers: Mapping[str, str] | None = None, **kwargs: Any
+    ) -> Response:
+        return HTTP.get_with_timeout(
+            url, headers=headers, auth=(self._username, self._password), **kwargs
+        )
+
+    @staticmethod
+    def _get_oauth_url_from_auth_document(
+        url: str, auth_document: Mapping[str, Any]
+    ) -> str:
+        auth_types: list[dict[str, Any]] = auth_document.get("authentication", [])
+        oauth_authentication = [
+            tlinks
+            for t in auth_types
+            if t.get("type") == "http://opds-spec.org/auth/oauth/client_credentials"
+            and (tlinks := t.get("links")) is not None
+        ]
+        if not oauth_authentication:
+            raise IntegrationException(
+                f"Unable to find supported authentication type ({url})",
+                debug_message=f"Auth document: {json.dumps(auth_document)}",
+            )
+
+        links = oauth_authentication[0]
+        auth_links: list[str] = [
+            lhref
+            for l in links
+            if l.get("rel") == "authenticate" and (lhref := l.get("href")) is not None
+        ]
+        if len(auth_links) != 1:
+            raise IntegrationException(
+                "Unable to find valid authentication link (%s)" % url,
+                debug_message=f"Found {len(auth_links)} authentication links. Auth document: {json.dumps(auth_document)}",
+            )
```

Comment:
Minor: This could be slightly misleading. Maybe we could adjust the wording to "...find exactly one valid..." make it a little more clear that we got zero or more than one.